napari/layers/image/experimental/octree.py

```diff
@@ -67,240 +74,285 @@ def _create_tiles(array: np.ndarray, tile_size: int) -> np.ndarray:
-    ordering which makes more sense, we should switch to it.
+        def _clamp(val, min_val, max_val):
+            return max(min(val, max_val), min_val)
 
-    -Z [0..3]
-    +Z [4..7]
+        tile_size = self.info.tile_size
 
-        -X X+
-    -Y 0 1
-    +Y 3 2
+        tiles = [span[0] / tile_size, span[1] / tile_size]
+        new_min = _clamp(tiles[0], 0, num_tiles - 1)
+        new_max = _clamp(tiles[1], 0, num_tiles - 1)
+        clamped = [new_min, new_max + 1]
 
-        -X X+
-    -Y 4 5
-    +Y 7 6
+        span_int = [int(x) for x in clamped]
+        return range(*span_int)
 
-    Parameters
-    ----------
-    row : int
-        The row of this octree node in its level.
-    col : int
-        The col of this octree node in its level.
-    data : np.ndarray
-        The image data for this octree node.
-    """
+    def row_range(self, span):
+        """Return row indices which span image coordinates [y0..y1]."""
+        return self.tile_range(span, self.num_rows)
 
-    def __init__(self, row: int, col: int, data: np.ndarray):
-        assert data is not None
-        assert row >= 0
-        assert col >= 0
-        self.row = row
-        self.col = col
-        self.data = data
-        self.children = None
+    def column_range(self, span):
+        """Return column indices which span image coordinates [x0..x1]."""
+        return self.tile_range(span, self.num_cols)
 
-    def print_info(self, level):
-        """Print information about this octree node.
 
-        level : int
-            The level of this node in the tree.
-        """
-        indent = "    " * level
-        print(
-            f"{indent}level={level} row={self.row:>3}, col={self.col:>3} "
-            f"shape={self.tile.shape}"
-        )
+def _one_tile(tiles: TileArray) -> bool:
+    return len(tiles) == 1 and len(tiles[0]) == 1
 
 
 class Octree:
-    """An octree.
+    """A region octree that holds hold 2D or 3D images.
+
+    Today the octree is full/complete meaning every node has 4 or 8
+    children, and every leaf node is at the same level of the tree. This
+    makes sense for region/image trees, because the image exists
+    everywhere.
+
+    Since we are a complete tree we don't need actual nodes with references
+    to the node's children. Instead, every level is just an array, and
+    going from parent to child or child to parent is trivial, you just
+    need to double or half the indexes.
+
+    Future Work: Geometry
+    ---------------------
+    Eventually we want our octree to hold geometry, not just images.
+    Geometry such as points and meshes. For geometry a sparse octree might
+    make more sense than this full/complete region octree.
+
+    With geometry there might be lots of empty space in between small dense
+    pockets of geometry. Some parts of tree might need to be very deep, but
+    it would be a waste for the tree to be that deep everywhere.
```

Comment:
Thanks for including this note

Reply:
The PCL octree is the best example of a fully developed octree for geometry. I didn't search far and wide but that's the most real-looking on I've seen:
https://pointclouds.org/documentation/group__octree.html